### PR TITLE
add dependencies variable to job script

### LINKED ISSUE #2555: Add information about prerequisites to task environment
I think it would be really useful to have task prerequisites available as a task environment variable. For example, this could used to work out task inputs based on its dependencies. I guess at the very minimum a list of `<task-id>:<state>` values would do, but for more complicated trigger conditions one might also be interested to know the actual states of the prerequisite tasks that fulfilled the trigger expression to work out which of them are relevant.

### PR DESCRIPTION
**Pull request by TomekTrzeciak**
Fixes #2555.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
This does not appear to give you the output/state?

**Comment by TomekTrzeciak** [comment]
Correct, this only gives satisfied dependencies as a list of task-ids, but I expect this should suffice in most use cases. For more detailed info I like your suggestion of `cylc show --json`.

**Comment by matthewrmshin** [comment]
Some initial comments:
* Need test, probably under this hierarchy https://github.com/cylc/cylc/tree/master/tests/jobscript
* Need to document the new environment variable. Somewhere here: https://github.com/cylc/cylc/blob/master/doc/src/cylc-user-guide/cug.tex#L4207
* What happens if we have a task that is triggered by a number of tasks?

**Comment by TomekTrzeciak** [comment]
Added test and docs. What happens for multiple triggers should be clear from the test (only the satisfied dependencies are listed).

File: tests/jobscript/18-env-task-dependencies.t [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+. "$(dirname "${0}")/test_header"
+set_test_number 3
+
+install_suite "${TEST_NAME_BASE}" "${TEST_NAME_BASE}"
+run_ok "${TEST_NAME_BASE}-validate" cylc validate "${SUITE_NAME}"
+run_ok "${TEST_NAME_BASE}-run" \
+    cylc run "${SUITE_NAME}" --reference-test --debug --no-detach
+grep_ok 'CYLC_TASK_DEPENDENCIES="bar\.1 baz\.1"' \
+    "${SUITE_RUN_DIR}/log/job/1/foo/01/job"
+
```
Comment (benfitzpatrick) [inline tests/jobscript/18-env-task-dependencies.t]: Add a grep for null `CYLC_TASK_DEPENDENCIES` in `qux/01/job`?

Comment (TomekTrzeciak) [reply tests/jobscript/18-env-task-dependencies.t]: Done

**Comment by hjoliver** [comment]
@matthewrmshin - given your comments above, can you do first review on this?

**Comment by matthewrmshin** [comment]
Looks good enough. What does it do if you manually trigger a downstream task that does not yet have its dependencies satisfied?

**Comment by TomekTrzeciak** [comment]
Hmm, I've just tried with manual trigger and it looks like the dependencies does not get properly updated. I used `tests/jobscript/18-env-task-dependencies/suite.rc` for testing (started in a held state and then triggering/resetting the state of different tasks).

From what I could figure out, the task has to go through the waiting state for its dependencies to get updated. Manual trigger doesn't seem to update the dependencies, it just uses the last updated state. Even more interestingly, by resetting the state of the task to ready , it gets its dependencies clear out entirely (empty CYLC_TASK_DEPENDENCIES variable). Not sure yet how to fix that.

**Comment by hjoliver** [comment]
Manual triggering gets a job to submit regardless of the state of its dependencies.

A forced reset to 'waiting' resets all dependencies to un-met (essentially the definition of waiting) so they can be re-met (if possible) in subsequent passes through the scheduling algorithm.

A forced reset to 'ready' is - I think - equivalent to 'all dependencies met' (leads to the same result as manual triggering, but less directly).  But its possible that the dependencies are not literally set to met in this case - we might just short-cut to the result of that.

**Comment by matthewrmshin** [comment]
(#2329 mentioned in offline discussion of this PR.)

**Comment by TomekTrzeciak** [comment]
After offline discussion and some further investigation I reckon the problem comes down to the fact that the current implementation of prerequisites doesn't directly reference the state of the task objects that they represent. As a result the state of prerequisite objects can diverge from the actual state of the tasks they represent. I guess fixing this goes beyond the scope of this PR and something like #2329 would need to happen first, before this PR could become viable. Feel free to close it for now.

**Comment by hjoliver** [comment]
@TomekTrzeciak - I agree that this can be done better (hence #2329) but note that forced triggering (and maybe forced state reset) seems to require that task and prerequisite states can be divorced sometimes. A manually triggered task typically does not have satisfied prerequisites (else it would not need manual triggering).

**Comment by hjoliver** [comment]
I'm not averse to merging this PR as-is if we document the limitations as discussed, if it would still be useful.  And/or would a list of all prerequisites (regardless of whether or not they are met) be useful in the task job environment?

**Comment by TomekTrzeciak** [comment]
I guess this PR could usable as is, but I just worried if it is usable enough for operational applications, where the suite operator != suite author and clear instructions must be given on how to handle failure states.

**Comment by TomekTrzeciak** [comment]
@hjoliver , but the motivating use case for this PR is exactly to make that problem easier, i.e., how to figure out (in a programmatically robust way) which tasks actually triggered the job.

Say, you have a graph like this:
```
FOO:succeed-any & FOO:finish-all => bar
```
and you want `bar` to know exactly which member(s) of `FOO` triggered it, not all the members of `FOO`.

(As a side note, I've checked the output of `cylc show` and for manually triggered tasks it reports that all prerequisites are met even for the tasks that have not run yet. Perhaps this comes down to the interpretation that if a task succeeded it automatically means that all its prerequisites are satisfied, but I find this rather misleading.)

**Comment by hjoliver** [comment]
> (As a side note, I've checked the output of cylc show and for manually triggered tasks it reports that all prerequisites are met even for the tasks that have not run yet. Perhaps this comes down to the interpretation that if a task succeeded it automatically means that all its prerequisites are satisfied, but I find this rather misleading.)

No, you've revealed a bug!  It used to be the case that `cylc show` listed individual components of prerequisite expressions (and the combined expression) as either satisfied or not.  

A quick manual bisect suggests this went wrong somewhere between 6.10.0 and 6.10.1 (I have not gone done to commit level yet).  Ouch.  I guess we don't have a test for this...

**Comment by hjoliver** [comment]
@TomekTrzeciak - after #2600 is merged, prerequisites should behave as you expect for this PR.

**Comment by TomekTrzeciak** [comment]
Brilliant!

**Comment by matthewrmshin** [comment]
I guess this one is now good to go?

**Comment by hjoliver** [comment]
(From a quick look, there's just one simple conflict).

**Comment by TomekTrzeciak** [comment]
I  rebased and fixed the conflict. One test failed, but it looks unrelated to this PR.

**Comment by hjoliver** [comment]
(restarted Travis CI)

**Pull request merged**